service/matching/version_rule_helpers.go

```diff
@@ -0,0 +1,244 @@
+	if force == true {
+		return nil
+	}
+	rules := slices.Clone(g.GetAssignmentRules())
+	if maxARs > 0 && len(rules) > maxARs {
+		// todo carly: should the max # of rules count rules that have been deleted but not yet cleaned up from the db?
+		return serviceerror.NewFailedPrecondition(fmt.Sprintf("update exceeds number of assignment rules permitted in namespace dynamic config (%v/%v)", len(rules), maxARs))
+	}
+	foundUnfiltered := false
+	for _, dbRule := range rules {
+		rule := dbRule.GetRule()
+		if dbRule.DeleteTimestamp == nil && isUnfiltered(rule) {
+			foundUnfiltered = true
+		}
+		tbid := rule.GetTargetBuildId()
+		if dbRule.DeleteTimestamp == nil && isInVersionSet(tbid, g.GetVersionSets()) {
+			return serviceerror.NewFailedPrecondition(fmt.Sprintf("update breaks requirement, target build id %s is already a member of version set", tbid))
+		}
+	}
+	if !foundUnfiltered {
+		return serviceerror.NewFailedPrecondition("update breaks requirement that at least one assignment rule must have no ramp or hint")
+	}
+	return nil
+}
+
+// isUnfiltered returns true if the assignment rule has no hint, no worker ratio ramp, and no ramp percentage, or a ramp percentage of 100
+func isUnfiltered(ar *taskqueue.BuildIdAssignmentRule) bool {
+	percentageRamp := ar.GetPercentageRamp()
+	return ar.GetHintFilter() == "" &&
+		ar.GetWorkerRatioRamp() == nil &&
+		(percentageRamp == nil || (percentageRamp != nil && percentageRamp.RampPercentage == 100))
+}
+
+// isInVersionSet returns true if the build id is in any of the listed version sets
+func isInVersionSet(tbid string, sets []*persistencepb.CompatibleVersionSet) bool {
+	for _, set := range sets {
+		for _, bid := range set.BuildIds {
+			if bid.GetId() == tbid {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func InsertAssignmentRule(timestamp *hlc.Clock,
+	data *persistencepb.VersioningData,
+	req *workflowservice.UpdateWorkerVersioningRulesRequest_InsertBuildIdAssignmentRule,
+	maxARs int) (*persistencepb.VersioningData, error) {
+	if data == nil {
+		data = &persistencepb.VersioningData{AssignmentRules: make([]*persistencepb.AssignmentRule, 0)}
+	} else {
+		data = common.CloneProto(data)
+	}
+	if req.GetRuleIndex() < 0 {
+		return nil, serviceerror.NewInvalidArgument("rule index cannot be negative")
+	}
+	persistenceAR := persistencepb.AssignmentRule{
+		Rule:            req.GetRule(),
+		CreateTimestamp: timestamp,
+		DeleteTimestamp: nil,
+	}
+	slices.Insert(data.GetAssignmentRules(), int(req.GetRuleIndex()), &persistenceAR)
+	return data, checkAssignmentConditions(data, maxARs, false)
+}
+
+func ReplaceAssignmentRule(timestamp *hlc.Clock,
+	data *persistencepb.VersioningData,
+	req *workflowservice.UpdateWorkerVersioningRulesRequest_ReplaceBuildIdAssignmentRule) (*persistencepb.VersioningData, error) {
+	if data == nil {
+		return nil, serviceerror.NewInvalidArgument("data cannot be nil for a replace call")
```

Comment:
since this error may be sent back to the user, I think it's better to let the code go forward in here so it can hit the out of bound error below. From user perspective we have an empty list not nil data.

Reply:
I think that `proto.Clone(data).(T)` will have a nil pointer error if we let it proceed forward, because `proto.Clone(data)` will be nil. I'll do your plan for now, and then add a test with nil data and see if it needs to be changed